File: shared/agent-core/src/heartbeat.ts
import type { TaskStore } from "./tasks";
import type { ForbiddenActions } from "./forbidden";
import type { AuditLog } from "./audit";
import type { GoalStore } from "./goals";
import type { AgentTask, ChannelMap, HeartbeatConfig, MonitorDefinition, PlatformAdapter, ProjectCode } from "./types";
import { MonitorSystem } from "./monitors";
import { selectIdleStrategy, buildIdlePrompt } from "./idle";

export interface HeartbeatDeps {
  taskStore: TaskStore;
  forbidden: ForbiddenActions;
  audit: AuditLog;
  platform: PlatformAdapter;
  config: HeartbeatConfig;
  notifyChatId: string;
  channelMap?: ChannelMap;
  monitors?: MonitorDefinition[];
  goalStore?: GoalStore;
  executeTask: (task: AgentTask) => Promise<{ result: string; costUsd?: number; durationMs?: number }>;
}

interface RunningTask {
  title: string;
  project: string;
  startedAt: number;
}

export class Heartbeat {
  private timer: ReturnType<typeof setInterval> | null = null;
  private stopped = false;
  private deps: HeartbeatDeps;

  // Per-project execution slots
  private runningTasks: Map<string, RunningTask> = new Map();

  // Idle tracking
  private lastTaskExecutedAt: number = Date.now();
  private idleTasksToday: number = 0;
  private idleTasksResetDate: string = new Date().toISOString().slice(0, 10);

  // Monitor system
  private monitorSystem: MonitorSystem | null = null;

  // Feature toggles (initialized from config)
  private idleEnabled: boolean;
  private chainingEnabled: boolean;
  private monitorsEnabled: boolean;

  constructor(deps: HeartbeatDeps) {
    this.deps = deps;
    this.idleEnabled = deps.config.idle?.enabled ?? false;
    this.chainingEnabled = deps.config.chainingEnabled ?? false;
    this.monitorsEnabled = deps.config.monitorsEnabled ?? false;
  }

  start(): void {
    this.stopped = false;
    console.log(
      `[heartbeat] Started (interval: ${this.deps.config.intervalMs / 1000}s)`
    );

    // Start monitor system if monitors provided and enabled
    if (this.monitorsEnabled && this.deps.monitors && this.deps.monitors.length > 0) {
      this.monitorSystem = new MonitorSystem(
        this.deps.monitors,
        (eventName, context) => this.fireEvent(eventName, context)
      );
      this.monitorSystem.start();
    }

    // Run first tick after a short delay
    setTimeout(() => this.tick(), 5000);

    this.timer = setInterval(
      () => this.tick(),
      this.deps.config.intervalMs
    );
  }

  stop(): void {
    this.stopped = true;
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
    if (this.monitorSystem) {
      this.monitorSystem.stop();
      this.monitorSystem = null;
    }
    console.log("[heartbeat] Stopped");
  }

  isStopped(): boolean {
    return this.stopped;
  }

  resume(): void {
    if (this.stopped) {
      this.start();
    }
  }

  private async tick(): Promise<void> {
    if (this.stopped) return;

    try {
      // Check quiet hours
      if (this.isQuietHours()) return;

      // Check daily budget
      const todayCost = await this.deps.audit.getTodayCost(
        this.deps.config.timezone
      );
      if (todayCost >= this.deps.config.dailyBudgetUsd) {
        console.log(
          `[heartbeat] Daily budget exceeded ($${todayCost.toFixed(2)} / $${this.deps.config.dailyBudgetUsd})`
        );
        return;
      }

      // Check rate limit
      const recentCount =
        await this.deps.audit.getProactiveCountLastHour();
      if (recentCount >= this.deps.config.maxProactivePerHour) {
        console.log(
          `[heartbeat] Rate limit reached (${recentCount}/${this.deps.config.maxProactivePerHour} per hour)`
        );
        return;
      }

      // Get due tasks
      const dueTasks = this.deps.taskStore.getDueTasks();

      if (dueTasks.length === 0) {
        // No due tasks — try idle tasks if enabled
        await this.maybeRunIdleTasks();
        return;
      }

      // Filter out tasks whose project already has a running task
      const tasksToRun = dueTasks.filter(t => {
        const key = t.project || "GENERAL";
        return !this.runningTasks.has(key);
      });

      if (tasksToRun.length === 0) {
        const running = Array.from(this.runningTasks.values()).map(r => r.project);
        console.log(`[heartbeat] All due tasks blocked by running: ${running.join(", ")}`);
        return;
      }

      console.log(`[heartbeat] ${tasksToRun.length} task(s) to run (${this.runningTasks.size} already running)`);

      // Launch all in parallel
      const promises = tasksToRun.map(task => this.executeTask(task));
      await Promise.allSettled(promises);
    } catch (err) {
      console.error("[heartbeat] Tick error:", err);
    }
  }

  private async executeTask(task: AgentTask): Promise<void> {
    const projectKey = task.project || "GENERAL";

    // Check if approval required
    if (task.requires_approval) {
      const targetChatId = this.getTargetChatId(task);
      await this.deps.platform.sendMessage(
        targetChatId,
        `[AUTO/${task.project}] 승인 필요: "${task.title}"\n실행하려면 /approve ${task.id.slice(0, 8)}`
      );
      return;
    }

    // Register running task
    this.runningTasks.set(projectKey, {
      title: task.title,
      project: projectKey,
      startedAt: Date.now(),
    });
    this.deps.taskStore.updateStatus(task.id, "running");
    console.log(`[heartbeat] Executing task: ${task.title} (${projectKey})`);

    try {
      const { result, costUsd, durationMs } =
        await this.deps.executeTask(task);

      // Update last task execution time
      this.lastTaskExecutedAt = Date.now();

      // Complete the run (handles recurring re-scheduling)
      this.deps.taskStore.completeRun(task.id, result, costUsd);

      // Audit
      await this.deps.audit.record({
        ts: new Date().toISOString(),
        type: "heartbeat",
        task: task.title,
        violations: [],
        cost: costUsd,
        duration: durationMs,
      });

      // Notify user if configured — route to project channel if mapped
      if (task.notify_user && result) {
        const targetChatId = this.getTargetChatId(task);
        const prefix = `[AUTO/${task.project}] `;
        const message =
          result.length > 3800
            ? prefix + result.slice(0, 3800) + "..."
            : prefix + result;
        await this.deps.platform.sendMessage(targetChatId, message);
      }

      console.log(
        `[heartbeat] Task "${task.title}" completed ($${costUsd?.toFixed(4) || "0"})`
      );
    } catch (err) {
      console.error(`[heartbeat] Task "${task.title}" failed:`, err);
      this.deps.taskStore.updateStatus(task.id, "failed");

      // Notify user of failure
      const targetChatId = this.getTargetChatId(task);
      await this.deps.platform.sendMessage(
        targetChatId,
        `[AUTO/${task.project}] 태스크 실패: "${task.title}"\n${String(err).slice(0, 200)}`
      );

      await this.deps.audit.record({
        ts: new Date().toISOString(),
        type: "heartbeat",
        task: task.title,
        violations: [],
      });
    } finally {
      this.runningTasks.delete(projectKey);
    }
  }

  /** Run idle tasks for projects that don't have a running task */
  private async maybeRunIdleTasks(): Promise<void> {
    if (!this.idleEnabled) return;

    const idleConfig = this.deps.config.idle;
    if (!idleConfig?.enabled) return;

    // Reset daily counter if date changed
    const today = new Date().toISOString().slice(0, 10);
    if (today !== this.idleTasksResetDate) {
      this.idleTasksToday = 0;
      this.idleTasksResetDate = today;
    }

    // Check daily idle task limit
    if (this.idleTasksToday >= idleConfig.maxIdleTasksPerDay) {
      return;
    }

    // Check idle threshold
    const elapsed = Date.now() - this.lastTaskExecutedAt;
    if (elapsed < idleConfig.idleThresholdMs) {
      return;
    }

    // Pick multiple idle strategies for different projects
    const projectPool: ProjectCode[] = ["COIN", "BLOG", "DASH", "TRAIN", "TGBOT", "DCBOT"];
    const available = projectPool.filter(p => !this.runningTasks.has(p));
    if (available.length === 0) return;

    // Run up to 3 idle tasks in parallel (respecting daily limit)
    const maxParallel = Math.min(3, available.length, idleConfig.maxIdleTasksPerDay - this.idleTasksToday);
    const idlePromises: Promise<void>[] = [];

    for (let i = 0; i < maxParallel; i++) {
      const { strategy, project } = selectIdleStrategy(undefined, this.deps.goalStore, available[i]);
      if (this.runningTasks.has(project)) continue;

      const prompt = buildIdlePrompt(strategy, project, this.deps.goalStore);
      console.log(`[heartbeat] Running idle task: ${strategy.title} (${project})`);

      const idleTask: AgentTask = {
        id: crypto.randomUUID(),
        type: "one-time",
        status: "running",
        title: `[IDLE] ${strategy.title} - ${project}`,
        prompt,
        project,
        schedule_cron: null,
        schedule_next: null,
        event_trigger: null,
        last_run_at: null,
        last_result: null,
        run_count: 0,
        max_runs: 1,
        notify_user: true,
        requires_approval: false,
        created_at: new Date().toISOString(),
        updated_at: new Date().toISOString(),
      };

      this.idleTasksToday++;
      idlePromises.push(this.executeIdleTask(idleTask, project));
    }

    await Promise.allSettled(idlePromises);
  }

  private async executeIdleTask(task: AgentTask, project: string): Promise<void> {
    const projectKey = project;
    this.runningTasks.set(projectKey, {
      title: task.title,
      project: projectKey,
      startedAt: Date.now(),
    });

    try {
      const { result, costUsd, durationMs } =
        await this.deps.executeTask(task);

      this.lastTaskExecutedAt = Date.now();

      // Audit
      await this.deps.audit.record({
        ts: new Date().toISOString(),
        type: "heartbeat",
        task: task.title,
        violations: [],
        cost: costUsd,
        duration: durationMs,
      });

      // Notify
      if (result) {
        const targetChatId = this.getTargetChatId(task);
        const prefix = `[IDLE/${project}] `;
        const message =
          result.length > 3800
            ? prefix + result.slice(0, 3800) + "..."
            : prefix + result;
        await this.deps.platform.sendMessage(targetChatId, message);
      }

      console.log(
        `[heartbeat] Idle task "${task.title}" completed ($${costUsd?.toFixed(4) || "0"})`
      );
    } catch (err) {
      console.error(`[heartbeat] Idle task "${task.title}" failed:`, err);
    } finally {
      this.runningTasks.delete(projectKey);
    }
  }

  /** Resolve the target chat/channel ID for a task based on its project */
  private getTargetChatId(task: AgentTask): string {
    if (this.deps.channelMap && task.project) {
      const mapped = this.deps.channelMap[task.project];
      if (mapped) return mapped;
    }
    return this.deps.notifyChatId;
  }

  private isQuietHours(): boolean {
    const { quietHoursStart, quietHoursEnd, timezone } = this.deps.config;
    const now = new Date();
    const hour = parseInt(
      now.toLocaleString("en-US", {
        timeZone: timezone,
        hour: "numeric",
        hour12: false,
      }),
      10
    );

    if (quietHoursStart > quietHoursEnd) {
      // e.g., 23:00 - 08:00 (crosses midnight)
      return hour >= quietHoursStart || hour < quietHoursEnd;
    }
    return hour >= quietHoursStart && hour < quietHoursEnd;
  }

  /** Fire event-triggered tasks, optionally with context */
  async fireEvent(eventName: string, context?: string): Promise<void> {
    if (this.stopped) return;

    // Guard: don't fire during quiet hours
    if (this.isQuietHours()) {
      console.log(`[heartbeat] Event "${eventName}" suppressed (quiet hours)`);
      return;
    }

    const tasks = this.deps.taskStore.getEventTasks(eventName);
    const promises = tasks.map(task => {
      if (context) {
        const enrichedTask = { ...task, prompt: `${task.prompt}\n\nContext: ${context}` };
        return this.executeTask(enrichedTask);
      }
      return this.executeTask(task);
    });
    await Promise.allSettled(promises);
  }

  // ─── Feature Toggles ───

  setIdleEnabled(enabled: boolean): void {
    this.idleEnabled = enabled;
    console.log(`[heartbeat] Idle tasks ${enabled ? "enabled" : "disabled"}`);
  }

  setChainingEnabled(enabled: boolean): void {
    this.chainingEnabled = enabled;
    console.log(`[heartbeat] Task chaining ${enabled ? "enabled" : "disabled"}`);
  }

  setMonitorsEnabled(enabled: boolean): void {
    this.monitorsEnabled = enabled;
    if (enabled && !this.monitorSystem && this.deps.monitors && this.deps.monitors.length > 0) {
      this.monitorSystem = new MonitorSystem(
        this.deps.monitors,
        (eventName, context) => this.fireEvent(eventName, context)
      );
      this.monitorSystem.start();
    } else if (!enabled && this.monitorSystem) {
      this.monitorSystem.stop();
      this.monitorSystem = null;
    }
    console.log(`[heartbeat] Monitors ${enabled ? "enabled" : "disabled"}`);
  }

  getMonitorSystem(): MonitorSystem | null {
    return this.monitorSystem;
  }

  isChainingEnabled(): boolean {
    return this.chainingEnabled;
  }

  // ─── Status ───

  async getStatus(): Promise<{
    running: boolean;
    runningTasks: RunningTask[];
    idleEnabled: boolean;
    chainingEnabled: boolean;
    monitorsEnabled: boolean;
    todayTaskCount: number;
    todayCost: number;
    lastTaskExecutedAt: number | null;
    monitorStatus: ReturnType<MonitorSystem["getStatus"]> | null;
  }> {
    const stats = await this.deps.audit.getTodayStats(this.deps.config.timezone);
    return {
      running: !this.stopped,
      runningTasks: Array.from(this.runningTasks.values()),
      idleEnabled: this.idleEnabled,
      chainingEnabled: this.chainingEnabled,
      monitorsEnabled: this.monitorsEnabled,
      todayTaskCount: stats.taskCount,
      todayCost: stats.totalCost,
      lastTaskExecutedAt: stats.lastTaskAt,
      monitorStatus: this.monitorSystem?.getStatus() ?? null,
    };
  }
}

File: content-pipeline/backend/src/agent/heartbeat.py
"""Heartbeat — main orchestrator. Port of agent-core/src/heartbeat.ts"""

import asyncio
import logging
import time
import uuid
from datetime import datetime

from src.agent.audit import AuditLog
from src.agent.claude_cli import call_claude
from src.agent.forbidden import ForbiddenActions
from src.agent.goals import GoalStore
from src.agent.idle import build_idle_prompt, select_idle_strategy
from src.agent.monitors import MonitorSystem
from src.agent.seed import seed_initial_tasks
from src.agent.soul import build_agent_system_prompt, load_soul
from src.agent.tags import process_tags
from src.agent.tasks import TaskStore
from src.agent.telegram import TelegramNotifier
from src.agent.types import AgentTask, MonitorDefinition
from src.config import Config

logger = logging.getLogger(__name__)

PROJECT_POOL = ["COIN", "BLOG", "DASH", "TRAIN", "TGBOT", "DCBOT"]

DEFAULT_MONITORS: list[MonitorDefinition] = [
    {
        "id": "health-all",
        "name": "Service Health Check",
        "event_name": "server_down",
        "interval_sec": 60,
        "enabled": True,
        "endpoints": [
            {"name": "coin-auto-trade", "url": "http://127.0.0.1:8001/health", "project": "COIN"},
            {"name": "train-go", "url": "http://127.0.0.1:8000/health", "project": "TRAIN"},
            {"name": "dashboard", "url": "http://127.0.0.1:8002/health", "project": "DASH"},
        ],
        "failure_threshold": 3,
    },
]


class RunningTask:
    def __init__(self, title: str, project: str):
        self.title = title
        self.project = project
        self.started_at = time.time()

    def to_dict(self) -> dict:
        return {
            "title": self.title,
            "project": self.project,
            "startedAt": int(self.started_at * 1000),
        }


class Heartbeat:
    def __init__(
        self,
        config: Config,
        task_store: TaskStore,
        audit: AuditLog,
        goal_store: GoalStore,
        forbidden: ForbiddenActions,
        notifier: TelegramNotifier,
    ):
        self.config = config
        self.task_store = task_store
        self.audit = audit
        self.goal_store = goal_store
        self.forbidden = forbidden
        self.notifier = notifier

        self._stopped = True
        self._task: asyncio.Task | None = None
        self._running_tasks: dict[str, RunningTask] = {}

        # Idle tracking
        self._last_task_at = time.time()
        self._idle_tasks_today = 0
        self._idle_reset_date = datetime.now().strftime("%Y-%m-%d")

        # Feature toggles (from agent_config DB)
        self._idle_enabled = False
        self._chaining_enabled = False
        self._monitors_enabled = False

        # Monitor system
        self._monitor_system: MonitorSystem | None = None

        # Soul content
        self._soul = ""

    async def start(self):
        self._stopped = False

        # Load soul
        if self.config.soul_path:
            self._soul = load_soul(self.config.soul_path)

        # Load forbidden
        if self.config.forbidden_config_path:
            self.forbidden.load()

        # Seed initial tasks
        seed_initial_tasks(self.task_store)

        # Recover zombie tasks
        self._recover_zombies()

        # Start monitors if enabled
        if self._monitors_enabled:
            self._start_monitors()

        logger.info(f"[heartbeat] Started (interval: {self.config.heartbeat_interval_sec}s)")

        self._task = asyncio.create_task(self._loop())

    def stop(self):
        self._stopped = True
        if self._task:
            self._task.cancel()
            self._task = None
        if self._monitor_system:
            self._monitor_system.stop()
            self._monitor_system = None
        logger.info("[heartbeat] Stopped")

    def is_stopped(self) -> bool:
        return self._stopped

    async def resume(self):
        if self._stopped:
            await self.start()

    def _recover_zombies(self):
        active = self.task_store.get_active()
        zombies = [
            t for t in active
            if t["status"] == "running" and (t.get("project") or "GENERAL") not in self._running_tasks
        ]
        for z in zombies:
            logger.info(f'[heartbeat] Recovering zombie task: "{z["title"]}" → pending')
            self.task_store.update_task(z["id"], {"status": "pending"})
        if zombies:
            logger.info(f"[heartbeat] Recovered {len(zombies)} zombie task(s)")

    def _start_monitors(self):
        if self._monitor_system:
            return
        self._monitor_system = MonitorSystem(
            DEFAULT_MONITORS,
            lambda event_name, context: self.fire_event(event_name, context),
        )
        self._monitor_system.start()

    async def _loop(self):
        # Initial delay
        await asyncio.sleep(5)

        while not self._stopped:
            try:
                await self._tick()
            except Exception as e:
                logger.error(f"[heartbeat] Tick error: {e}")
            await asyncio.sleep(self.config.heartbeat_interval_sec)

    async def _tick(self):
        if self._stopped:
            return

        # Periodically recover zombies
        self._recover_zombies()

        # Check quiet hours
        if self._is_quiet_hours():
            return

        # Check daily budget
        today_cost = self.audit.get_today_cost(self.config.user_timezone)
        if today_cost >= self.config.agent_daily_budget_usd:
            logger.info(f"[heartbeat] Daily budget exceeded (${today_cost:.2f} / ${self.config.agent_daily_budget_usd})")
            return

        # Check rate limit
        recent_count = self.audit.get_proactive_count_last_hour()
        if recent_count >= self.config.max_proactive_per_hour:
            logger.info(f"[heartbeat] Rate limit reached ({recent_count}/{self.config.max_proactive_per_hour} per hour)")
            return

        # Get due tasks
        due_tasks = self.task_store.get_due_tasks()

        if not due_tasks:
            await self._maybe_run_idle()
            return

        # Filter tasks already running for the same project
        tasks_to_run = [
            t for t in due_tasks
            if (t.get("project") or "GENERAL") not in self._running_tasks
        ]

        if not tasks_to_run:
            running = [r.project for r in self._running_tasks.values()]
            logger.info(f"[heartbeat] All due tasks blocked by running: {', '.join(running)}")
            return

        logger.info(f"[heartbeat] {len(tasks_to_run)} task(s) to run ({len(self._running_tasks)} already running)")

        # Launch all in parallel
        results = await asyncio.gather(
            *[self._execute_task(t) for t in tasks_to_run],
            return_exceptions=True,
        )
        for r in results:
            if isinstance(r, Exception):
                logger.error(f"[heartbeat] Task execution exception: {r}")

    async def _execute_task(self, task: AgentTask):
        project_key = task.get("project") or "GENERAL"

        # Check if approval required
        if task["requires_approval"]:
            await self.notifier.send_message(
                f'[AUTO/{task["project"]}] 승인 필요: "{task["title"]}"\n'
                f'실행하려면 /approve {task["id"][:8]}',
            )
            return

        # Register running task
        self._running_tasks[project_key] = RunningTask(task["title"], project_key)
        self.task_store.update_status(task["id"], "running")
        logger.info(f'[heartbeat] Executing task: {task["title"]} ({project_key})')

        try:
            result_data = await self._call_claude_for_task(task)
            result_text = result_data["result"]
            cost_usd = result_data.get("cost_usd")
            duration_ms = result_data.get("duration_ms")

            # Update last task time
            self._last_task_at = time.time()

            # Complete the run
            self.task_store.complete_run(task["id"], result_text, cost_usd)

            # Audit
            self.audit.record(
                ts=datetime.now().isoformat(),
                audit_type="heartbeat",
                task=task["title"],
                cost=cost_usd,
                duration=duration_ms,
            )

            # Notify user
            if task["notify_user"] and result_text:
                prefix = f'[AUTO/{task["project"]}] '
                message = prefix + (result_text[:3800] + "..." if len(result_text) > 3800 else result_text)
                await self.notifier.send_message(message)

            logger.info(f'[heartbeat] Task "{task["title"]}" completed (${cost_usd or 0:.4f})')

        except Exception as e:
            logger.error(f'[heartbeat] Task "{task["title"]}" failed: {e}')
            self.task_store.update_status(task["id"], "failed")

            await self.notifier.send_message(
                f'[AUTO/{task["project"]}] 태스크 실패: "{task["title"]}"\n{str(e)[:200]}'
            )

            self.audit.record(
                ts=datetime.now().isoformat(),
                audit_type="heartbeat",
                task=task["title"],
            )
        finally:
            self._running_tasks.pop(project_key, None)

    async def _call_claude_for_task(self, task: AgentTask) -> dict:
        time_str = datetime.now().strftime("%A, %B %d, %Y %I:%M %p")

        goals_context = ""
        if self.goal_store:
            goals = self.goal_store.get_by_project(task.get("project", "GENERAL"))
            if goals:
                goals_context = "\n".join(
                    f"- [{g['priority'].upper()}] {g['title']}"
                    + (f" (공유: {', '.join(g['projects'])})" if len(g["projects"]) > 1 else "")
                    for g in goals
                )

        active_tasks = self.task_store.get_active()
        tasks_summary = "\n".join(f"- {t['title']} ({t['type']})" for t in active_tasks)

        system_prompt = build_agent_system_prompt(
            soul=self._soul,
            forbidden_block=self.forbidden.format_for_prompt(),
            active_tasks_summary=tasks_summary,
            user_name=self.config.user_name,
            current_time=time_str,
            goals_context=goals_context,
            chaining_enabled=self._chaining_enabled,
        )

        session_id = str(uuid.uuid4())

        # Progress callback: send to Telegram
        def on_progress(msg: str):
            asyncio.create_task(self.notifier.send_message(
                f'[AUTO/{task.get("project", "GENERAL")}] {msg}'
            ))

        result = await call_claude(
            prompt=task["prompt"],
            session_id=session_id,
            is_new_session=True,
            system_prompt=system_prompt,
            claude_path=self.config.claude_path,
            cwd=self.config.project_dir,
            on_progress=on_progress,
        )

        if result["success"]:
            # Process tags
            tag_result = process_tags(
                result["result"],
                self.task_store,
                chain_depth=task.get("chain_depth", 0),
            )
            return {
                "result": tag_result.clean_text,
                "cost_usd": result.get("cost_usd"),
                "duration_ms": result.get("duration_ms"),
            }

        return {
            "result": result.get("error") or "Task failed",
            "cost_usd": result.get("cost_usd"),
            "duration_ms": result.get("duration_ms"),
        }

    async def _maybe_run_idle(self):
        if not self._idle_enabled:
            return

        # Reset daily counter
        today = datetime.now().strftime("%Y-%m-%d")
        if today != self._idle_reset_date:
            self._idle_tasks_today = 0
            self._idle_reset_date = today

        # Check daily limit
        if self._idle_tasks_today >= self.config.idle_max_per_day:
            return

        # Check idle threshold
        elapsed = time.time() - self._last_task_at
        if elapsed < self.config.idle_threshold_sec:
            return

        # Find available projects (not currently running)
        available = [p for p in PROJECT_POOL if p not in self._running_tasks]
        if not available:
            return

        # Run up to 3 idle tasks in parallel
        max_parallel = min(3, len(available), self.config.idle_max_per_day - self._idle_tasks_today)
        idle_coros = []

        for i in range(max_parallel):
            strategy, project = select_idle_strategy(
                goal_store=self.goal_store, project=available[i]
            )
            if project in self._running_tasks:
                continue

            prompt = build_idle_prompt(strategy, project, self.goal_store)
            logger.info(f"[heartbeat] Running idle task: {strategy.title} ({project})")

            idle_task: AgentTask = {
                "id": str(uuid.uuid4()),
                "type": "one-time",
                "status": "running",
                "title": f"[IDLE] {strategy.title} - {project}",
                "prompt": prompt,
                "project": project,
                "schedule_cron": None,
                "schedule_next": None,
                "event_trigger": None,
                "last_run_at": None,
                "last_result": None,
                "run_count": 0,
                "max_runs": 1,
                "notify_user": True,
                "requires_approval": False,
                "chain_depth": 0,
                "chain_parent_id": None,
                "created_at": datetime.now().isoformat(),
                "updated_at": datetime.now().isoformat(),
            }

            self._idle_tasks_today += 1
            idle_coros.append(self._execute_idle_task(idle_task, project))

        await asyncio.gather(*idle_coros, return_exceptions=True)

    async def _execute_idle_task(self, task: AgentTask, project: str):
        self._running_tasks[project] = RunningTask(task["title"], project)

        try:
            result_data = await self._call_claude_for_task(task)
            result_text = result_data["result"]
            cost_usd = result_data.get("cost_usd")
            duration_ms = result_data.get("duration_ms")

            self._last_task_at = time.time()

            self.audit.record(
                ts=datetime.now().isoformat(),
                audit_type="heartbeat",
                task=task["title"],
                cost=cost_usd,
                duration=duration_ms,
            )

            if result_text:
                prefix = f"[IDLE/{project}] "
                message = prefix + (result_text[:3800] + "..." if len(result_text) > 3800 else result_text)
                await self.notifier.send_message(message)

            logger.info(f'[heartbeat] Idle task "{task["title"]}" completed (${cost_usd or 0:.4f})')

        except Exception as e:
            logger.error(f'[heartbeat] Idle task "{task["title"]}" failed: {e}')
        finally:
            self._running_tasks.pop(project, None)

    def _is_quiet_hours(self) -> bool:
        start = self.config.quiet_hours_start
        end = self.config.quiet_hours_end
        if start < 0 or end < 0:
            return False

        now = datetime.now()
        hour = now.hour

        if start > end:
            return hour >= start or hour < end
        return start <= hour < end

    async def fire_event(self, event_name: str, context: str | None = None):
        if self._stopped:
            return
        if self._is_quiet_hours():
            logger.info(f'[heartbeat] Event "{event_name}" suppressed (quiet hours)')
            return

        tasks = self.task_store.get_event_tasks(event_name)
        coros = []
        for task in tasks:
            if context:
                enriched = dict(task)
                enriched["prompt"] = f"{task['prompt']}\n\nContext: {context}"
                coros.append(self._execute_task(enriched))
            else:
                coros.append(self._execute_task(task))
        await asyncio.gather(*coros, return_exceptions=True)

    # ─── Feature Toggles ───

    def set_idle_enabled(self, enabled: bool):
        self._idle_enabled = enabled
        logger.info(f"[heartbeat] Idle tasks {'enabled' if enabled else 'disabled'}")

    def set_chaining_enabled(self, enabled: bool):
        self._chaining_enabled = enabled
        logger.info(f"[heartbeat] Task chaining {'enabled' if enabled else 'disabled'}")

    def set_monitors_enabled(self, enabled: bool):
        self._monitors_enabled = enabled
        if enabled and not self._monitor_system:
            self._start_monitors()
        elif not enabled and self._monitor_system:
            self._monitor_system.stop()
            self._monitor_system = None
        logger.info(f"[heartbeat] Monitors {'enabled' if enabled else 'disabled'}")

    # ─── Status ───

    async def get_status(self) -> dict:
        stats = self.audit.get_today_stats(self.config.user_timezone)
        return {
            "running": not self._stopped,
            "runningTasks": [r.to_dict() for r in self._running_tasks.values()],
            "idleEnabled": self._idle_enabled,
            "chainingEnabled": self._chaining_enabled,
            "monitorsEnabled": self._monitors_enabled,
            "todayTaskCount": stats["task_count"],
            "todayCost": stats["total_cost"],
            "lastTaskExecutedAt": stats["last_task_at"],
            "monitorStatus": self._monitor_system.get_status() if self._monitor_system else None,
        }
